File: src/services/calendar.js
const { google } = require('googleapis');
const pool = require('../db/pool');
const { createOAuthClient } = require('./googleAuth');

// Builds an OAuth2 client pre-loaded with this user's refresh token.
// The googleapis library uses it to silently mint a fresh access token
// before each API call — we never handle that refresh step ourselves.
async function getAuthorizedClientForUser(userId) {
  const { rows } = await pool.query(
    'SELECT google_refresh_token FROM users WHERE id = $1',
    [userId]
  );
  const refreshToken = rows[0]?.google_refresh_token;

  if (!refreshToken) {
    throw new Error(`User ${userId} has not connected Google Calendar yet`);
  }

  const oauth2Client = createOAuthClient();
  oauth2Client.setCredentials({ refresh_token: refreshToken });
  return oauth2Client;
}

async function listUpcomingEvents(userId, maxResults = 10) {
  const auth = await getAuthorizedClientForUser(userId);
  const calendar = google.calendar({ version: 'v3', auth });

  const { data } = await calendar.events.list({
    calendarId: 'primary',
    timeMin: new Date().toISOString(),
    maxResults,
    singleEvents: true, // expand recurring events into individual instances
    orderBy: 'startTime',
  });

  return data.items;
}

async function createEvent(userId, { summary, description, startDateTime, endDateTime, timeZone }) {
  const auth = await getAuthorizedClientForUser(userId);
  const calendar = google.calendar({ version: 'v3', auth });

  const { data } = await calendar.events.insert({
    calendarId: 'primary',
    requestBody: {
      summary,
      description,
      start: { dateTime: startDateTime, timeZone },
      end: { dateTime: endDateTime, timeZone },
    },
  });

  return data;
}

module.exports = { listUpcomingEvents, createEvent };

File: app/services/calendar.py
from datetime import datetime, timezone
from typing import Optional

from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from sqlalchemy.orm import Session

from app.config import settings
from app.db.models import User
from app.services.token_crypto import decrypt_token


# Builds a Credentials object holding only a refresh token. The
# google-api-python-client transport layer sees it has no valid access
# token and silently exchanges the refresh token for one before the
# first API call — we never handle that refresh step ourselves.
def _get_credentials_for_user(db: Session, user_id: int) -> Credentials:
    user = db.get(User, user_id)
    if not user or not user.google_refresh_token:
        raise ValueError(f"User {user_id} has not connected Google Calendar yet")

    return Credentials(
        token=None,
        refresh_token=decrypt_token(user.google_refresh_token),
        token_uri="https://oauth2.googleapis.com/token",
        client_id=settings.google_client_id,
        client_secret=settings.google_client_secret,
    )


def list_upcoming_events(db: Session, user_id: int, max_results: int = 10):
    creds = _get_credentials_for_user(db, user_id)
    service = build("calendar", "v3", credentials=creds)

    now = datetime.now(timezone.utc).isoformat()
    result = (
        service.events()
        .list(
            calendarId="primary",
            timeMin=now,
            maxResults=max_results,
            singleEvents=True,  # expand recurring events into individual instances
            orderBy="startTime",
        )
        .execute()
    )

    return result.get("items", [])


def create_event(
    db: Session,
    user_id: int,
    summary: str,
    description: str,
    start_datetime: str,
    end_datetime: str,
    time_zone: str,
    recurrence: Optional[list[str]] = None,
):
    creds = _get_credentials_for_user(db, user_id)
    service = build("calendar", "v3", credentials=creds)

    event_body = {
        "summary": summary,
        "description": description,
        "start": {"dateTime": start_datetime, "timeZone": time_zone},
        "end": {"dateTime": end_datetime, "timeZone": time_zone},
    }
    if recurrence:
        event_body["recurrence"] = recurrence  # e.g. ["RRULE:FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR"]

    return service.events().insert(calendarId="primary", body=event_body).execute()
